src/FlatSharp.Compiler/UnityNativeArrayVectorTypeModel.cs

```diff
@@ -0,0 +1,66 @@
+using FlatSharp.CodeGen;
+using FlatSharp.Internal;
+using FlatSharp.TypeModel;
+using Unity.Collections;
+
+namespace FlatSharp.Compiler;
+
+#if NETSTANDARD2_1 || NET5_0_OR_GREATER
+
+/// <summary>
+/// Defines a vector type model for a Unity NativeArray collection.
+/// </summary>
+public class UnityNativeArrayVectorTypeModel : BaseVectorTypeModel
+{
+    internal UnityNativeArrayVectorTypeModel(Type vectorType, TypeModelContainer provider) : base(vectorType, provider)
+    {
+    }
+
+    public override string LengthPropertyName => nameof(Memory<byte>.Length);
+
+    public override Type OnInitialize()
+    {
+        return this.ClrType.GetGenericArguments()[0];
+    }
+
+    protected override string CreateLoop(FlatBufferSerializerOptions options, string vectorVariableName, string numberOfItemsVariableName, string expectedVariableName, string body)
+    {
+        FlatSharpInternal.Assert(false, "Not expecting to do loop get max size for memory vector");
+        throw new Exception();
+    }
+
+    public override CodeGeneratedMethod CreateParseMethodBody(ParserCodeGenContext context)
+    {
+        ValidateWriteThrough(
+            writeThroughSupported: false,
+            this,
+            context.AllFieldContexts,
+            context.Options);
+
+        var readNativeArray = $"{context.InputBufferVariableName}.{nameof(UnityExtensions.ReadNativeArrayBlock)}<{this.ItemTypeModel.GetGlobalCompilableTypeName()}, {context.InputBufferTypeName}>({context.OffsetVariableName})";
+
+        return new CodeGeneratedMethod($"return {readNativeArray};");
+    }
+
+    public override CodeGeneratedMethod CreateSerializeMethodBody(SerializationCodeGenContext context)
+    {
+        var writeNativeArray = $"{context.SpanWriterVariableName}.{nameof(SpanWriterUnityExtensions.WriteNativeArrayBlock)}({context.SpanVariableName}, {context.ValueVariableName}, {context.OffsetVariableName}, {context.SerializationContextVariableName});";
+
+        return new CodeGeneratedMethod(writeNativeArray);
+    }
+
+    public override CodeGeneratedMethod CreateCloneMethodBody(CloneCodeGenContext context)
+    {
+        string body = $"throw new NotSupportedException();";
```

Comment:
this will break copy constructors that the compiler generates. Since it fails predictably, that seems not-terrible, but it's going to be a non-obvious failure. Is there a case where NativeArrays should be cloneable? Since they seem immutable it might be enough to just do the equivalent of what we do for strings and primitives where we just return the original value and the whole thing becomes a no-op.

Reply:
They are mutable. I am looking into handling some sort of Clone now.